Source/i_glob.c

```diff
@@ -0,0 +1,376 @@
+//
+// Copyright(C) 2018 Simon Howard
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+//
+// File globbing API. This allows the contents of the filesystem
+// to be interrogated.
+//
+
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+
+#include "i_glob.h"
+#include "m_misc.h"
+#include "m_misc2.h"
+#include "config.h"
+
+#if defined(_MSC_VER)
+// For Visual C++, we need to include the win_opendir module.
+#include "win_opendir.h"
+#define strcasecmp _stricmp
+#define strncasecmp _strnicmp
+#include <sys/stat.h>
+#define S_ISDIR(m)      (((m)& S_IFMT) == S_IFDIR)
+#elif defined(HAVE_DIRENT_H)
+#include <dirent.h>
+#include <sys/stat.h>
+#elif defined(__WATCOMC__)
+// Watcom has the same API in a different header.
+#include <direct.h>
+#else
+#define NO_DIRENT_IMPLEMENTATION
+#endif
+
+#ifndef NO_DIRENT_IMPLEMENTATION
+
+// Only the fields d_name and (as an XSI extension) d_ino are specified
+// in POSIX.1.  Other than Linux, the d_type field is available mainly
+// only on BSD systems.  The remaining fields are available on many, but
+// not all systems.
+static boolean IsDirectory(char *dir, struct dirent *de)
+{
+#if defined(_DIRENT_HAVE_D_TYPE)
+    if (de->d_type != DT_UNKNOWN && de->d_type != DT_LNK)
+    {
+        return de->d_type == DT_DIR;
+    }
+    else
+#endif
+    {
+        char *filename;
+        struct stat sb;
+        int result;
+
+        filename = M_StringJoin(dir, DIR_SEPARATOR_S, de->d_name, NULL);
+        result = stat(filename, &sb);
+        free(filename);
```

Comment:
I've got `Z_Free: freed pointer without ZONEID` here. By the way, do you recommend using zone memory in general? I used it in `u_mapinfo.c` and I'm not sure.

Reply:
Sigh, thank you! I guess it's obvious now that this was a blind commit without a chance to test the resulting executable. :wink: 

Regarding zone memory: I tend to use plain `malloc()` for everything that's not part of the game data, i.e. for everything that doesn't benefit from zone memory's ability to "cache" data instead of outright `free()`ing it. Most non-pristine additions to the port do not fall into this category, though, thus I prefer POSIX ` malloc()`.